Mono.Debugging/Mono.Debugging.Evaluation/AsyncOperationBase.cs

```diff
@@ -0,0 +1,88 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Mono.Debugging.Client;
+
+namespace Mono.Debugging.Evaluation
+{
+	public class OperationResult<TValue>
+	{
+		public TValue Result { get; private set; }
+		public bool ResultIsException { get; private set; }
+
+		public OperationResult (TValue result, bool resultIsException)
+		{
+			Result = result;
+			ResultIsException = resultIsException;
+		}
+	}
+
+	public static class OperationResultEx
+	{
+		public static OperationResult<TValue> ThrowIfException<TValue> (this OperationResult<TValue> result, EvaluationContext ctx)
+		{
+			if (!result.ResultIsException)
+				return result;
+			var exceptionTypeName = ctx.Adapter.GetValueTypeName (ctx, result.Result);
+			throw new EvaluatorExceptionThrownException (result.Result, exceptionTypeName);
+		}
+	}
+
+	public interface IAsyncOperationBase
+	{
+		Task RawTask { get; }
+		string Description { get; }
+		void AfterCancelled (int elapsedAfterCancelMs);
```

Comment:
It's used in CorDebug (monodevelop repo). It's done to ensure that we awaited enough amount of time after Abort() call. Otherwise we can get exceptions in CorDebug.